File: 1014-Best-Sightseeing-Pair/best_sightseeing_pair.py
# You are given an integer array values where values[i] represents the value of the ith sightseeing spot.
# Two sightseeing spots i and j have a distance j - i between them.
# The score of a pair (i < j) of sightseeing spots is values[i] + values[j] + i - j:
# the sum of the values of the sightseeing spots, minus the distance between them.
# Return the maximum score of a pair of sightseeing spots.

from typing import List

class Solution:
    def maxScoreSightseeingPair(self, values: List[int]) -> int:
        max_value = 0
        array = []
        for i in range(len(values)):
            pair_sum = values[i] + i
            if pair_sum > max_value:
                max_value = pair_sum
                array.append(max_value)
            else:
                array.append(max_value)

        max_score = 0
        for i in range(len(values) - 1, 0, -1):
            pair_sum = values[i] - i
            if pair_sum + array[i - 1] > max_score:
                max_score = pair_sum + array[i - 1]
        return max_score

if __name__ == "__main__":
    object = Solution()

    print(object.maxScoreSightseeingPair([8, 1, 5, 2, 6]))
    print(object.maxScoreSightseeingPair([1, 2]))
    print(object.maxScoreSightseeingPair([8, 1, 5, 2, 13, 6, 3, 7, 12]))

File: Leetcode/1014-Best-Sightseeing-Pair/best_sightseeing_pair.cpp
#include<iostream>
#include<vector>
using namespace std;

class Solution{
    public:
    int maxScoreSightseeingPair(vector<int>& values) {
        int max_value = 0;
        int pair_sum;
        vector<int>array;
        for(int i=0; i<values.size(); i++){
            pair_sum = values.at(i) + i;
            if (pair_sum > max_value){
                max_value = pair_sum;
                array.push_back(max_value);
            }
            else{array.push_back(max_value);}
        }
        int max_score = 0;
        for (int i=values.size()-1; i>0; i--){
            pair_sum = values.at(i) - i;
            if (pair_sum + array.at(i-1)>max_score){
                max_score = pair_sum + array.at(i-1);
            }
        }
        return max_score;
    }
};
int main(){
    Solution Sol;
    vector<int> vec1 = {8,1,5,2,6};
    vector<int> vec2 = {1,2};
    cout << Sol.maxScoreSightseeingPair(vec1) << endl;
    cout << Sol.maxScoreSightseeingPair(vec2) << endl;
}